### PR TITLE
Explain why 'elision'

### PR DESCRIPTION
**Pull request by steveklabnik**

### CODE REVIEW

**Comment by Gankro** [comment]
I'm not clear if there's a rigorous distinction between elision and inference. Both will put different values based on context. Inference fails out if it's ambiguous, elision fails out if what is picked doesn't work? Inference is also more best-effort, while elision is about certain if-then rules?

**Comment by steveklabnik** [comment]
@Gankro to me, elision is a passive thing, while inference is an active thing.

**Comment by Kimundi** [comment]
Well, technically lifetime elision is a very restricted secondary type inference algorithm that only applies in item signatures, where the primary one is forbidden. That is to say, Rust might be a bit weird in that regard because it has more than one type inference algorithm. (Other languages that claim to infer types have also different algorithms, but I'm not aware of a case where there are multiple ones)

It probably makes sense to keep telling them apart as "inference" vs "ellision" as a shorthand, but the fact that both automatically infer some types without needing to write them down should not be glossed over.

The best example here is that `&T` in a function signature uses the ellision-inference algorithm to figure out what `'a` to pick, while `&T` in the body of the function use the local-inference algorithm.

---

For the docs, maybe an approach that makes both the distinction and similarity clear would work?

"Rust supports powerful local type inference in function bodies, but it's forbidden in item signatures to allow reasoning about the types just based in the item signature alone. However, for ergonomic reasons a very restricted secondary inference algorithm called "lifetime elision" applies in function signatures. It infers only based on the signature components themselves and not based on the body of the function, only infers lifetime paramters, and does this with only three easily memorizable and unambiguous rules. This makes lifetime elision a shorthand for writing an item signature, while not hiding away the actual types involved as the full local inference would if applied to it.

**Comment by steveklabnik** [comment]
What do you think of that version, @Gankro ?

**Comment by Gankro** [comment]
:+1: that's a great explanation, @Kimundi!

**Comment by steveklabnik** [comment]
@Gankro updated to use @Kimundi 's wording. r?

File: src/doc/trpl/ownership.md [hunk 1]
```diff
@@ -472,10 +472,16 @@ thread-safe counterpart of `Rc<T>`.
 
 ## Lifetime Elision
 
-Earlier, we mentioned *lifetime elision*, a feature of Rust which allows you to
-not write lifetime annotations in certain circumstances. All references have a
-lifetime, and so if you elide a lifetime (like `&T` instead of `&'a T`), Rust
-will do three things to determine what those lifetimes should be.
+Rust supports powerful local type inference in function bodies, but it’s
+forbidden in item signatures to allow reasoning about the types just based in
```
Comment (Gankro) [inline src/doc/trpl/ownership.md]: s/in/on?

File: src/doc/trpl/ownership.md [hunk 2]
```diff
@@ -472,10 +472,16 @@ thread-safe counterpart of `Rc<T>`.
...
+the item signature alone. However, for ergonomic reasons a very restricted
+secondary inference algorithm called “lifetime elision” applies in function
+signatures. It infers only based on the signature components themselves and not
+based on the body of the function, only infers lifetime paramters, and does
+this with only three easily memorizable and unambiguous rules. This makes
+lifetime elision a shorthand for writing an item signature, while not hiding
+away the actual types involved as the full local inference would if applied to
```
Comment (Gankro) [inline src/doc/trpl/ownership.md]: Drop "the" before full local inference?

File: src/doc/trpl/ownership.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/trpl/ownership.md]: I would say 'in' here, but can see 'on' too.

**Comment by Gankro** [comment]
r=me with nits

File: src/doc/trpl/ownership.md (hunk 2 above)
Comment (steveklabnik) [inline src/doc/trpl/ownership.md]: also reads fine to me this way, but yours as well

**Comment by steveklabnik** [comment]
I agree with one nit, but disagree with the other. :)

**Pull request merged**